modmail/extensions/threads.py

```diff
@@ -0,0 +1,459 @@
+import asyncio
+import datetime
+import logging
+from typing import TYPE_CHECKING, Optional, Union
+
+import discord
+from arrow import Arrow
+from discord.ext import commands, tasks
+from discord.ext.commands import Context
+from discord.utils import escape_markdown
+
+from modmail.utils.cogs import ExtMetadata, ModmailCog
+from modmail.utils.converters import Duration
+from modmail.utils.threads import ThreadEmbed, Ticket, is_modmail_thread
+from modmail.utils.threads.errors import ThreadAlreadyExistsError, ThreadNotFoundError
+from modmail.utils.users import check_can_dm_user
+
+
+if TYPE_CHECKING:
+    from modmail.bot import ModmailBot
+    from modmail.log import ModmailLogger
+
+EXT_METADATA = ExtMetadata()
+
+
+BASE_JUMP_URL = "https://discord.com/channels"
+USER_NOT_ABLE_TO_BE_DMED_MESSAGE = (
+    "**{0}** is not able to be dmed! This is because they have either blocked the bot, "
+    "or they are only accepting direct messages from friends.\n"
+    "It is also possible that they are not in the same server as the bot. "
+)
+ON_SUCCESS_EMOJI = "\u2705"  # ✅
+
+# This will be part of configuration later, so its stored in globals for now
+ENABLE_DM_TO_GUILD_TYPING = False  # Library bug prevents this form working right now
+ENABLE_GUILD_TO_DM_TYPING = False
+
+USE_AUDIT_LOGS = True
+logger: "ModmailLogger" = logging.getLogger(__name__)
+
+
+class TicketsCog(ModmailCog, name="Threads"):
+    """A cog for relaying direct messages."""
+
+    def __init__(self, bot: "ModmailBot"):
+        self.bot = bot
+        # user id, Ticket
+        self.relay_channel: Union[
+            discord.TextChannel, discord.PartialMessageable
+        ] = self.bot.get_partial_messageable(self.bot.config.thread.relay_channel_id)
```

Comment:
We should validate this, to make the user has supplied a discord text channel and not a thread.

Reply:
Will defer to a separate pr, since validation must also occur when setting the configuration item.